pkg/proxy/proxy.go

```diff
@@ -1,267 +0,0 @@
-func New(client clientset.Interface, config Config, authorizer authorizer.Authorizer, authenticator authenticator.Request) (*kubeRBACProxy, error) {
-	return new(authenticator, authorizer, config), nil
-}
-
-// Handle authenticates the client and authorizes the request.
-// If the authn fails, a 401 error is returned. If the authz fails, a 403 error is returned
-func (h *kubeRBACProxy) Handle(w http.ResponseWriter, req *http.Request) bool {
-	ctx := req.Context()
-	if len(h.Config.Authentication.Token.Audiences) > 0 {
-		ctx = authenticator.WithAudiences(ctx, h.Config.Authentication.Token.Audiences)
-		req = req.WithContext(ctx)
-	}
-
-	// Authenticate
-	u, ok, err := h.AuthenticateRequest(req)
-	if err != nil {
-		klog.Errorf("Unable to authenticate the request due to an error: %v", err)
-		http.Error(w, "Unauthorized", http.StatusUnauthorized)
-		return false
-	}
-	if !ok {
-		http.Error(w, "Unauthorized", http.StatusUnauthorized)
-		return false
-	}
-
-	// Get authorization attributes
-	allAttrs := h.authorizerAttributesGetter.GetRequestAttributes(u.User, req)
-	if len(allAttrs) == 0 {
-		msg := "Bad Request. The request or configuration is malformed."
-		klog.V(2).Info(msg)
-		http.Error(w, msg, http.StatusBadRequest)
-		return false
-	}
-
-	for _, attrs := range allAttrs {
-		// Authorize
-		authorized, reason, err := h.Authorize(ctx, attrs)
-		if err != nil {
-			msg := fmt.Sprintf("Authorization error (user=%s, verb=%s, resource=%s, subresource=%s)", u.User.GetName(), attrs.GetVerb(), attrs.GetResource(), attrs.GetSubresource())
-			klog.Errorf("%s: %s", msg, err)
-			http.Error(w, msg, http.StatusInternalServerError)
-			return false
-		}
-		if authorized != authorizer.DecisionAllow {
-			msg := fmt.Sprintf("Forbidden (user=%s, verb=%s, resource=%s, subresource=%s)", u.User.GetName(), attrs.GetVerb(), attrs.GetResource(), attrs.GetSubresource())
-			klog.V(2).Infof("%s. Reason: %q.", msg, reason)
-			http.Error(w, msg, http.StatusForbidden)
-			return false
-		}
-	}
-
-	if h.Config.Authentication.Header.Enabled {
-		// Seemingly well-known headers to tell the upstream about user's identity
-		// so that the upstream can achieve the original goal of delegating RBAC authn/authz to kube-rbac-proxy
-		headerCfg := h.Config.Authentication.Header
-		req.Header.Set(headerCfg.UserFieldName, u.User.GetName())
-		req.Header.Set(headerCfg.GroupsFieldName, strings.Join(u.User.GetGroups(), headerCfg.GroupSeparator))
-	}
-
-	return true
-}
-
-func newKubeRBACProxyAuthorizerAttributesGetter(authzConfig *authz.Config) *krpAuthorizerAttributesGetter {
-	return &krpAuthorizerAttributesGetter{authzConfig}
-}
-
-type krpAuthorizerAttributesGetter struct {
-	authzConfig *authz.Config
-}
-
-// GetRequestAttributes populates authorizer attributes for the requests to kube-rbac-proxy.
-func (n krpAuthorizerAttributesGetter) GetRequestAttributes(u user.Info, r *http.Request) []authorizer.Attributes {
-	apiVerb := ""
-	switch r.Method {
```

Comment:
Falling through with an empty verb isn't what I would expect.  Why does it do this instead of `Unknown_r.Method` or failing?

Reply:
TODO: pre-acceptance closing the empty verb hole

Reply:
didn't find good precedence in k8s which quickly looking but we should investigate.

Reply:
If the method is not specified, SAR returns allowed only if the identity is allowed to use all verbs with `"*"`, but we could make it explicit by starting of with `"*"` instead of `""`.